src/miner.cpp

```diff
@@ -305,6 +331,26 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
             if (!ContextualCheckInputs(tx, state, view, true, MANDATORY_SCRIPT_VERIFY_FLAGS, true, txdata, Params().GetConsensus(), consensusBranchId))
                 continue;
 
+            if (this_is_testnet_or_regtest && monitoring_pool_balances) {
+                // Does this transaction lead to a turnstile violation?
+
+                CAmount sproutValueDummy = sproutValue;
+                CAmount saplingValueDummy = saplingValue;
+
+                sproutValueDummy += -tx.valueBalance;
+
+                for (auto js : tx.vjoinsplit) {
+                    sproutValueDummy += js.vpub_old;
+                    sproutValueDummy -= js.vpub_new;
+                }
+
+                if (sproutValueDummy < 0) continue;
```

Comment:
Given the significance, maybe we should log this event under a new logging category e.g. "turnstile"?  For audit purposes, we should log the entire transaction rather than just the txid, since the transaction will be evicted from the mempool.

Reply:
Good idea, I'll make it log. However, transaction won't be evicted from the mempool. This code will just avoid mining it.